llvm/lib/CodeGen/WindowScheduler.cpp

```diff
@@ -0,0 +1,692 @@
+}
+
+void WindowScheduler::restoreTripleMBB() {
+  // After list scheduling, the MBB is restored in one traversal.
+  for (size_t I = 0; I < TriMIs.size(); ++I) {
+    auto *MI = TriMIs[I];
+    auto OldPos = MBB->begin();
+    std::advance(OldPos, I);
+    auto CurPos = MI->getIterator();
+    if (CurPos != OldPos) {
+      MBB->splice(OldPos, MBB, CurPos);
+      Context->LIS->handleMove(*MI, /*UpdateFlags=*/false);
+    }
+  }
+}
+
+SmallVector<unsigned> WindowScheduler::getSearchIndexes(unsigned SearchNum,
+                                                        unsigned SearchRatio) {
+  // We use SearchRatio to get the index range, and then evenly get the indexes
+  // according to the SearchNum. This is a simple huristic. Depending on the
+  // characteristics of the target, more complex algorithms can be used for both
+  // performance and compilation time.
+  assert(SearchRatio <= 100 && "SearchRatio should be equal or less than 100!");
+  unsigned MaxIdx = SchedInstrNum * SearchRatio / 100;
+  unsigned Step = SearchNum > 0 && SearchNum <= MaxIdx ? MaxIdx / SearchNum : 1;
+  SmallVector<unsigned> SearchIndexes;
+  for (unsigned Idx = 0; Idx < MaxIdx; Idx += Step)
+    SearchIndexes.push_back(Idx);
+  return SearchIndexes;
+}
+
+int WindowScheduler::getEstimatedII(ScheduleDAGInstrs &DAG) {
+  // Sometimes MaxDepth is 0, so it should be limited to the minimum of 1.
+  unsigned MaxDepth = 1;
+  for (auto &SU : DAG.SUnits)
+    MaxDepth = std::max(SU.getDepth() + SU.Latency, MaxDepth);
+  return MaxDepth * WindowIICoeff;
+}
+
+int WindowScheduler::calculateMaxCycle(ScheduleDAGInstrs &DAG,
+                                       unsigned Offset) {
+  int InitII = getEstimatedII(DAG);
+  ResourceManager RM(Subtarget, &DAG);
+  RM.init(InitII);
+  // ResourceManager and DAG are used to calculate the maximum cycle for the
+  // scheduled MIs. Since MIs in the Region have already been scheduled, the
+  // emit cycles can be estimated in order here.
+  int CurCycle = 0;
+  auto Range = getScheduleRange(Offset, SchedInstrNum);
+  for (auto &MI : Range) {
+    auto *SU = DAG.getSUnit(&MI);
+    int ExpectCycle = CurCycle;
+    // The predecessors of current MI determine its earliest issue cycle.
+    for (auto &Pred : SU->Preds) {
+      auto *PredMI = Pred.getSUnit()->getInstr();
+      int PredCycle = getOriCycle(PredMI);
+      ExpectCycle = std::max(ExpectCycle, PredCycle + (int)Pred.getLatency());
+    }
+    // ResourceManager can be used to detect resource conflicts between the
+    // current MI and the previously inserted MIs.
+    while (!RM.canReserveResources(*SU, CurCycle) || CurCycle < ExpectCycle) {
+      ++CurCycle;
+      if (CurCycle == (int)WindowIILimit)
+        return CurCycle;
+    }
+    RM.reserveResources(*SU, CurCycle);
+    OriToCycle[getOriMI(&MI)] = CurCycle;
+    LLVM_DEBUG(dbgs() << "\tCycle " << CurCycle << " [S."
+                      << getOriStage(getOriMI(&MI), Offset) << "]: ");
+    LLVM_DEBUG(MI.dump());
```

Comment:
Can just directly << MI in the first debug statement

Reply:
Updated